### ISSUE TITLE
PE level 6 dropSword

### ISSUE THREAD

**Issue by zhiqitan** [nus-cs2030/2021-s1#340]
<img width="635" alt="Screenshot 2020-11-13 at 8 05 39 PM" src="https://user-images.githubusercontent.com/67494720/99070896-146b0680-25ec-11eb-9be4-93978231c9cf.png">
<img width="535" alt="Screenshot 2020-11-13 at 8 05 46 PM" src="https://user-images.githubusercontent.com/67494720/99070919-1cc34180-25ec-11eb-90bd-90e040960a0d.png">

hi, any hints to handle dropSword? I keep having this output, but not sure where goes wrong. Thanks in advance for all the replies!

**Comment by domlimm** [comment]
Hi, consider this:

In the current room, if you have a Sword, you need to bring it along when you return to the previous room. (Something to take note of too!)

EDIT: Zhi Qi, apologies, I may not have answered your question correctly. Please look at Ernest's and Friscilia's instead!

**Comment by ErnestCYW** [comment]
I think it is because when you drop the sword it does not return a new Room with the dropped sword inside it, what your program is doing is probably just calling the previous room or just printing you dropped the sword without reflecting it in your room.

Maybe you can add something in your sword to show if it's dropped or not? Or you can also try returning a completely new Room with the same name but previous rooms things + the dropped sword?

**Comment by frisciliasultan** [comment]
Hi, I think you might have removed the sword from the room when you call dropSword. Consider an attribute in your sword that keeps track of the status of the sword whether it is taken or not, instead of removing the sword completely so that you still hv the line "sword is shimmering". Hope it helps!

**Comment by YuFeng0930** [comment]
Hi, 
You can consider having a boolean value in your Sword class to help you keep track of whether the sword is taken or not. DropSword simply replaces the old sword with a new Sword with the correct boolean value in the room, not completely deleting the sword in the room. The only time you need to delete the sword from the room is when you go to the next room, delete the sword in the previous room (which will appear in the later test case). Other than that, keep your sword in the room.

**Comment by zhiqitan** [comment]
> Hi, consider this:
> 
> In the current room, if you have a Sword, you need to bring it along when you return to the previous room. (Something to take note of too!)
> 
> EDIT: Zhi Qi, apologies, I may not have answered your question correctly. Please look at Ernest's and Friscilia's instead!

hi, no worries. Thanks for willing to help!!

**Comment by zhiqitan** [comment]
Thanks @frisciliasultan  and @YuFeng0930 for the ideas. It works but now having issues for test6_1.

**Comment by zhiqitan** [comment]
<img width="1190" alt="Screenshot 2020-11-14 at 1 13 36 AM" src="https://user-images.githubusercontent.com/67494720/99103143-1481fb00-261a-11eb-9def-ed104805c45b.png">
Anyone has the ideas for the problems? Looks like I have the issue of excessive swords and the sequence of the items. How can I handle the exception for no room to back?

**Comment by hhhhanz** [comment]
I believe if the room has no prev room, it should return this

**Comment by hhhhanz** [comment]
I think for the excessive Swords, could be an issue with the list of items returned in your back() method. Maybe can share some pseudocode regarding how your back() works?

**Comment by zhiqitan** [comment]
@hhhhanz hi, I created a new room for the previous room and I check if current room has taken sword, if yes then I check if previous room has sword, if no then I add new taken sword, or if previous room has sword but was not taken then i remove the not taken sword from the previous room and add new taken sword to the previous room at behind. (is it I suppose to just change the previous room sword to taken state so the first issue will be solved? not sure for this)

**Comment by hhhhanz** [comment]
Not sure why you have to do this: if previous room has sword but was not taken then i remove the not taken sword from the previous room and add new taken sword to the previous room at behind. => why must go 2 rooms behind? (Sorry if I'm not understanding you correctly!)

To my understanding, as long as 
1) current room has sword
2) you are holding this sword
3) previous room has no sword 
then you move the taken sword to the prev room.

**Comment by YuFeng0930** [comment]
Try not to create a new room for the previous room, as the previous room should already exist. You can try to have a variable preRoom to store your previous room (it works for me). This is more like what the game actually is, instead of creating a new room for the previous room.

**Comment by zhiqitan** [comment]
@hhhhanz sorry for confuse u, I mean add new taken sword at the back of the items list of previous room. For your case, what if the previous room has sword but not taken, when I bring the taken sword back to previous room, should the sword added to the back of item list of the previous room?

**Comment by YuFeng0930** [comment]
For me, I do not add it back to the previous room. In my code, my room always has at most one sword. 

Besides, I may misinterpret your question in my last comment. What I want to say in my previous comment is just that you need not use the keyword ```new``` to create a new room for the previous room.

**Comment by zhiqitan** [comment]
@YuFeng0930 ya, understood, I have amended it to no new. but now I still have the first problem, not sure what's the test case

**Comment by YuFeng0930** [comment]
Do you mean the issue of excessive swords? If yes, then try this: If the sword is taken with you when you ```go``` to the next room, delete the sword in the previous room. Since the sword is taken to the new room with you, the previous copy of the sword should be deleted. (This change is in ```go()```, not in ```back()```)

**Comment by zhiqitan** [comment]
@YuFeng0930 yeah, I just realised it after thinking of what u saying about should think it like a game ! Thanks a lot for your help!!
